docs/website/docs/developers/debugging/llvm.md

```diff
@@ -0,0 +1,136 @@
+
+## Reducing optimization levels in IREE
+
+If you suspect an LLVM bug, try disabling (or reducing) one or both optimization
+levels. LLVM has two places where a `-O[n]` is applied: the middle-end
+(`opt`) and the backend/codegen (`llc`). The backend optimization level
+is selected by values of the `llvm::CodeGenOptLevel` enum, which is passed to
+a `createTargetMachine` call in the compiler plugin. This level defaults to
+`-O3`. On the other hand, the generic/middle-end `opt` optimization level
+is controlled by `llvm::OptimizationLevel` and defaults to `-O2` currently.
+
+Setting one or both of these values to the `-O0` or `-O1` equivalent
+and seeing the issue go away is an indicator that there may be a LLVM bug
+in play. It may, however, also indicate that there's a race condition or
+other correctness issue in the generated LLVM IR that is masked by a lack
+of compiler optimizations.
+
+## Useful flags
+
+- `-print-after-all` and `-print-before/after=[passname]` cane help
+  locate places were suspect IR is introduced or where crashes occur, just as their
+  MLIR equivalents can be used in IREE.
+- `-print-module-scope` ensures IR dumps include attributes and metadata
+  if those are relevant
+- The exact process for feeding a binary back into IREE after manually compiling
+  it is target-specific, but will generally involve
+  `--iree-hal-substitute-executable-object=[executable]=[filename]`.
+- `-global-isel=1` can be helpful in localizing a bug to instruction
+  selection. If it solves your problem (or turns it into a different bug),
+  you've substantially narrowed down the code that needs to be searched.
+- `opt` produces human-readable output whin passed the `-S` flag, and often
+  needs a `-o -` to send its results to standard output. `llc` takes a
+  `--filetype={asm,obj}` argument to control whether assembly or
+  assembled objects are produced.
+
+## `llvm-diff`
+
+When adjusting an `opt` invocation to isolate misbehaving passes or when
+comparing LLVM IR from a working and a broken commit, you may be able to
+use the `llvm-diff` tool to compare two LLVM IR files without the noise that
+is induced by LLVM's IR numbering scheme. Note that the diff is generall output
+to standard error and should be redirected with a `2>&1`.
+
+## `llvm-reduce`
+
+In some cases - particularly compiler crashes, the `llvm-reduce` program
+(part of LLVM) may be useful. It takes a LLVM IR file and an "interestingness
+script" which returns 0 (success) if there **is** a problem with a proposed
+reduced input and fails otherwise.
+
+When writing such a script, the `not` tool (especially its
+`not --crash [program] [args]` mode) and `FileCheck` from the LLVM test suite
+are often useful.
+
+[This is a helpful LLVM slide deck on how to operate `llvm-reduce`](https://www.llvm.org/devmtg/2025-04/slides/tutorial/arsenault_reduce.pdf).
+These slides include other useful flags and tips.
+
+## Creating reproducers
```

Comment:
A few concrete examples could be useful here, either inlined commands to run or links to issues / discussion threads where this reproducer process has been followed with success